File: lib/catalog/relink.ts
import {
  parseAssetId,
  parseCatalogId,
  parseOperationId,
  parseRootId,
  type AssetId,
  type CatalogId,
  type OperationId,
  type RootId,
} from "./ids.ts";
import { parseRelativePath, parseSessionId, type SessionId } from "./runtime.ts";

export type RelinkMatchRank =
  | "exact-sha256"
  | "local-file-identity"
  | "relative-observation"
  | "filename-size";

export type RelinkFingerprintStatus = "missing" | "hashing" | "valid" | "stale" | "failed";

export interface RelinkObservation {
  readonly byteLength: number | null;
  readonly modifiedAt: number | null;
  readonly localFileId: string | null;
}

export interface RelinkFingerprint {
  readonly status: RelinkFingerprintStatus;
  readonly sha256: string | null;
}

export interface RelinkMissingAsset {
  readonly assetId: AssetId;
  readonly rootId: RootId;
  readonly relativePath: string;
  readonly filename: string;
  readonly observation: RelinkObservation | null;
  readonly fingerprint: RelinkFingerprint;
}

export interface RelinkCandidate {
  readonly candidateId: string;
  readonly rootId: RootId;
  readonly relativePath: string;
  readonly filename: string;
  readonly observation: RelinkObservation;
  readonly fingerprint: RelinkFingerprint;
}

export interface RelinkPlanInput {
  readonly catalogId: CatalogId;
  readonly sessionId: SessionId;
  readonly operationId: OperationId;
  readonly missingAssets: readonly RelinkMissingAsset[];
  readonly candidates: readonly RelinkCandidate[];
}

export interface RelinkPair {
  readonly assetId: AssetId;
  readonly candidateId: string;
  readonly rank: RelinkMatchRank;
}

export interface RelinkSuggestion {
  readonly assetId: AssetId;
  readonly rank: RelinkMatchRank | null;
  readonly candidateIds: readonly string[];
  readonly preselectedCandidateId: string | null;
}

export interface RelinkDraft {
  readonly catalogId: CatalogId;
  readonly sessionId: SessionId;
  readonly operationId: OperationId;
  readonly missingAssets: readonly RelinkMissingAsset[];
  readonly candidates: readonly RelinkCandidate[];
  readonly suggestions: readonly RelinkSuggestion[];
  readonly preselectedPairs: readonly RelinkPair[];
  readonly unresolvedAssetCount: number;
  readonly ambiguousAssetCount: number;
}

export interface RelinkAcceptedPairInput {
  readonly assetId: AssetId;
  readonly candidateId: string;
}

export interface RelinkApplyResult {
  readonly catalogId: CatalogId;
  readonly sessionId: SessionId;
  readonly operationId: OperationId;
  readonly acceptedPairs: readonly RelinkPair[];
  readonly unresolvedAssetIds: readonly AssetId[];
  readonly unresolvedAssetCount: number;
  readonly unresolvedCandidateCount: number;
  readonly ambiguousAssetCount: number;
}

export const RELINK_MAX_ACCEPTED_PAIRS = 250;

export interface RelinkPrepareRequest {
  readonly catalogId: CatalogId;
  readonly sessionId: SessionId;
}

export interface RelinkApplyRequest {
  readonly catalogId: CatalogId;
  readonly sessionId: SessionId;
  readonly operationId: OperationId;
  readonly acceptedPairs: readonly RelinkAcceptedPairInput[];
}

export interface RelinkCancelRequest {
  readonly catalogId: CatalogId;
  readonly sessionId: SessionId;
  readonly operationId: OperationId;
}

export interface RelinkServiceDraft extends RelinkDraft {
  readonly expiresAt: number;
}

export interface RelinkServiceApplyResult extends RelinkApplyResult {
  readonly revision: number;
  readonly changed: boolean;
  readonly appliedMutations: number;
  readonly auditId: number | null;
}

type RecordValue = Record<string, unknown>;

const FINGERPRINT_STATUSES: readonly RelinkFingerprintStatus[] = [
  "missing",
  "hashing",
  "valid",
  "stale",
  "failed",
];

function recordValue(value: unknown, name: string): RecordValue {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`${name} must be an object.`);
  }
  return value as RecordValue;
}

function assertKnownKeys(
  input: RecordValue,
  allowed: readonly string[],
  name: string,
): void {
  for (const key of Object.keys(input)) {
    if (!allowed.includes(key)) throw new Error(`${name} contains unsupported fields.`);
  }
}

function stringValue(value: unknown, name: string, allowEmpty = false): string {
  if (
    typeof value !== "string" ||
    (!allowEmpty && value.length === 0) ||
    value.length > 512 ||
    value.includes("\0")
  ) {
    throw new Error(`${name} is invalid.`);
  }
  return value;
}

function candidateId(value: unknown): string {
  const parsed = stringValue(value, "candidateId");
  if (
    parsed.includes("/") ||
    parsed.includes("\\") ||
    parsed.startsWith(".") ||
    /^[A-Za-z]:/.test(parsed)
  ) {
    throw new Error("candidateId must not contain a native path.");
  }
  return parsed;
}

function filename(value: unknown): string {
  const parsed = stringValue(value, "filename");
  if (parsed === "." || parsed === ".." || parsed.includes("/") || parsed.includes("\\")) {
    throw new Error("filename is invalid.");
  }
  return parsed;
}

function numberValue(value: unknown, name: string, integer = false): number {
  if (
    typeof value !== "number" ||
    !Number.isFinite(value) ||
    (integer && !Number.isSafeInteger(value)) ||
    value < 0
  ) {
    throw new Error(`${name} is invalid.`);
  }
  return value;
}

function nullableString(value: unknown, name: string): string | null {
  return value === null ? null : stringValue(value, name, true);
}

function parseObservation(value: unknown, name: string): RelinkObservation | null {
  if (value === null) return null;
  const input = recordValue(value, name);
  return {
    byteLength: input.byteLength === null ? null : numberValue(input.byteLength, `${name}.byteLength`, true),
    modifiedAt: input.modifiedAt === null ? null : numberValue(input.modifiedAt, `${name}.modifiedAt`),
    localFileId: nullableString(input.localFileId, `${name}.localFileId`),
  };
}

function parseFingerprint(value: unknown, name: string): RelinkFingerprint {
  const input = recordValue(value, name);
  const status = FINGERPRINT_STATUSES.find((item) => item === input.status);
  if (status === undefined) {
    throw new Error(`${name}.status is invalid.`);
  }
  const sha256 = nullableString(input.sha256, `${name}.sha256`);
  if (sha256 !== null && !/^[0-9a-f]{64}$/.test(sha256)) {
    throw new Error(`${name}.sha256 is invalid.`);
  }
  if ((status === "valid") !== (sha256 !== null)) {
    throw new Error(`${name} status and sha256 do not agree.`);
  }
  return { status, sha256 };
}

function parseMissingAsset(value: unknown): RelinkMissingAsset {
  const input = recordValue(value, "missing asset");
  const relativePath = parseRelativePath(input.relativePath);
  const parsedFilename = filename(input.filename);
  if (relativePath.split("/").at(-1) !== parsedFilename) {
    throw new Error("missing asset filename does not match relativePath.");
  }
  return {
    assetId: parseAssetId(input.assetId),
    rootId: parseRootId(input.rootId),
    relativePath,
    filename: parsedFilename,
    observation: parseObservation(input.observation, "missing asset.observation"),
    fingerprint: parseFingerprint(input.fingerprint, "missing asset.fingerprint"),
  };
}

function parseCandidate(value: unknown): RelinkCandidate {
  const input = recordValue(value, "relink candidate");
  const relativePath = parseRelativePath(input.relativePath);
  const parsedFilename = filename(input.filename);
  if (relativePath.split("/").at(-1) !== parsedFilename) {
    throw new Error("candidate filename does not match relativePath.");
  }
  return {
    candidateId: candidateId(input.candidateId),
    rootId: parseRootId(input.rootId),
    relativePath,
    filename: parsedFilename,
    observation: (input.observation === undefined
      ? null
      : parseObservation(input.observation, "candidate.observation")) ?? {
      byteLength: null,
      modifiedAt: null,
      localFileId: null,
    },
    fingerprint: parseFingerprint(input.fingerprint, "candidate.fingerprint"),
  };
}

function parseArray(value: unknown, name: string): readonly unknown[] {
  if (!Array.isArray(value)) throw new Error(`${name} must be an array.`);
  return value;
}

function parsePlanInput(value: unknown): RelinkPlanInput {
  const input = recordValue(value, "relink plan");
  const missingAssets = parseArray(input.missingAssets, "missingAssets").map(parseMissingAsset);
  const candidates = parseArray(input.candidates, "candidates").map(parseCandidate);
  const assetIds = new Set<string>();
  for (const asset of missingAssets) {
    if (assetIds.has(asset.assetId)) throw new Error("missingAssets contains a duplicate asset.");
    assetIds.add(asset.assetId);
  }
  const candidateIds = new Set<string>();
  for (const candidate of candidates) {
    if (candidateIds.has(candidate.candidateId)) throw new Error("candidates contains a duplicate candidate.");
    candidateIds.add(candidate.candidateId);
  }
  return {
    catalogId: parseCatalogId(input.catalogId),
    sessionId: parseSessionId(input.sessionId),
    operationId: parseOperationId(input.operationId),
    missingAssets,
    candidates,
  };
}

function validHash(fingerprint: RelinkFingerprint): string | null {
  return fingerprint.status === "valid" ? fingerprint.sha256 : null;
}

function localIdentityMatches(asset: RelinkMissingAsset, candidate: RelinkCandidate): boolean {
  const assetId = asset.observation?.localFileId;
  return assetId !== null && assetId !== undefined && assetId !== "" && assetId === candidate.observation.localFileId;
}

function compatibleObservation(asset: RelinkMissingAsset, candidate: RelinkCandidate): boolean {
  const observation = asset.observation;
  return (
    observation !== null &&
    observation.byteLength !== null &&
    observation.modifiedAt !== null &&
    candidate.observation.byteLength === observation.byteLength &&
    candidate.observation.modifiedAt === observation.modifiedAt
  );
}

function filenameSizeMatches(asset: RelinkMissingAsset, candidate: RelinkCandidate): boolean {
  return (
    asset.filename === candidate.filename &&
    asset.observation?.byteLength !== null &&
    asset.observation?.byteLength !== undefined &&
    candidate.observation.byteLength === asset.observation.byteLength
  );
}

function candidateIds(candidates: readonly RelinkCandidate[]): readonly string[] {
  return candidates.map((candidate) => candidate.candidateId);
}

function suggestionFor(
  asset: RelinkMissingAsset,
  candidates: readonly RelinkCandidate[],
  usedCandidateIds: ReadonlySet<string>,
): RelinkSuggestion {
  const available = candidates.filter((candidate) => !usedCandidateIds.has(candidate.candidateId));
  const hash = validHash(asset.fingerprint);
  const exactHash =
    hash === null
      ? []
      : available.filter((candidate) => validHash(candidate.fingerprint) === hash);
  if (exactHash.length > 0) {
    return {
      assetId: asset.assetId,
      rank: "exact-sha256",
      candidateIds: candidateIds(exactHash),
      preselectedCandidateId: null,
    };
  }
  const local = available.filter((candidate) => localIdentityMatches(asset, candidate));
  if (local.length > 0) {
    return {
      assetId: asset.assetId,
      rank: "local-file-identity",
      candidateIds: candidateIds(local),
      preselectedCandidateId: null,
    };
  }
  const relative = available.filter(
    (candidate) => candidate.relativePath === asset.relativePath && compatibleObservation(asset, candidate),
  );
  if (relative.length > 0) {
    return {
      assetId: asset.assetId,
      rank: "relative-observation",
      candidateIds: candidateIds(relative),
      preselectedCandidateId: null,
    };
  }
  const filenameMatches = available.filter((candidate) => filenameSizeMatches(asset, candidate));
  return {
    assetId: asset.assetId,
    rank: filenameMatches.length > 0 ? "filename-size" : null,
    candidateIds: candidateIds(filenameMatches),
    preselectedCandidateId: null,
  };
}

export function planRelink(value: unknown): RelinkDraft {
  const input = parsePlanInput(value);
  const exactMatches = new Map<string, readonly RelinkCandidate[]>();
  const exactOwners = new Map<string, number>();
  for (const asset of input.missingAssets) {
    const hash = validHash(asset.fingerprint);
    const matches =
      hash === null
        ? []
        : input.candidates.filter((candidate) => validHash(candidate.fingerprint) === hash);
    exactMatches.set(asset.assetId, matches);
    for (const candidate of matches) {
      exactOwners.set(candidate.candidateId, (exactOwners.get(candidate.candidateId) ?? 0) + 1);
    }
  }

  const preselectedPairs: RelinkPair[] = [];
  const usedCandidateIds = new Set<string>();
  for (const asset of input.missingAssets) {
    const matches = exactMatches.get(asset.assetId) ?? [];
    if (matches.length === 1 && exactOwners.get(matches[0]!.candidateId) === 1) {
      const candidate = matches[0]!;
      preselectedPairs.push({ assetId: asset.assetId, candidateId: candidate.candidateId, rank: "exact-sha256" });
      usedCandidateIds.add(candidate.candidateId);
    }
  }

  const suggestions = input.missingAssets.map((asset) => {
    const preselected = preselectedPairs.find((pair) => pair.assetId === asset.assetId);
    if (preselected !== undefined) {
      return {
        assetId: asset.assetId,
        rank: "exact-sha256" as const,
        candidateIds: [preselected.candidateId],
        preselectedCandidateId: preselected.candidateId,
      };
    }
    const exact = exactMatches.get(asset.assetId) ?? [];
    if (exact.length > 0) {
      return {
        assetId: asset.assetId,
        rank: "exact-sha256" as const,
        candidateIds: candidateIds(exact),
        preselectedCandidateId: null,
      };
    }
    return suggestionFor(asset, input.candidates, usedCandidateIds);
  });
  const ambiguousAssetCount = suggestions.filter(
    (suggestion) => suggestion.preselectedCandidateId === null && suggestion.candidateIds.length > 1,
  ).length;
  return {
    catalogId: input.catalogId,
    sessionId: input.sessionId,
    operationId: input.operationId,
    missingAssets: input.missingAssets,
    candidates: input.candidates,
    suggestions,
    preselectedPairs,
    unresolvedAssetCount: input.missingAssets.length - preselectedPairs.length,
    ambiguousAssetCount,
  };
}

function parseAcceptedPairs(value: unknown): readonly RelinkAcceptedPairInput[] {
  const acceptedPairs = parseArray(value, "acceptedPairs").map((item) => {
    const input = recordValue(item, "accepted pair");
    assertKnownKeys(input, ["assetId", "candidateId"], "Accepted pair");
    return { assetId: parseAssetId(input.assetId), candidateId: candidateId(input.candidateId) };
  });
  if (acceptedPairs.length > RELINK_MAX_ACCEPTED_PAIRS) {
    throw new Error("acceptedPairs is too large.");
  }
  return acceptedPairs;
}

export function parseRelinkPrepareRequest(value: unknown): RelinkPrepareRequest {
  const input = recordValue(value, "relink prepare request");
  assertKnownKeys(input, ["catalogId", "sessionId"], "Relink prepare request");
  return {
    catalogId: parseCatalogId(input.catalogId),
    sessionId: parseSessionId(input.sessionId),
  };
}

export function parseRelinkApplyRequest(value: unknown): RelinkApplyRequest {
  const input = recordValue(value, "relink apply request");
  assertKnownKeys(input, ["catalogId", "sessionId", "operationId", "acceptedPairs"], "Relink apply request");
  return {
    catalogId: parseCatalogId(input.catalogId),
    sessionId: parseSessionId(input.sessionId),
    operationId: parseOperationId(input.operationId),
    acceptedPairs: parseAcceptedPairs(input.acceptedPairs),
  };
}

export function parseRelinkCancelRequest(value: unknown): RelinkCancelRequest {
  const input = recordValue(value, "relink cancel request");
  assertKnownKeys(input, ["catalogId", "sessionId", "operationId"], "Relink cancel request");
  return {
    catalogId: parseCatalogId(input.catalogId),
    sessionId: parseSessionId(input.sessionId),
    operationId: parseOperationId(input.operationId),
  };
}

export function applyRelinkDraft(
  draftValue: unknown,
  acceptedValue: unknown,
): RelinkApplyResult {
  const draft = planRelink(draftValue);
  const requested = parseAcceptedPairs(acceptedValue);
  const suggestionByAsset = new Map(draft.suggestions.map((suggestion) => [suggestion.assetId, suggestion]));
  const acceptedPairs = [...draft.preselectedPairs];
  const acceptedAssetIds = new Set(acceptedPairs.map((pair) => pair.assetId));
  const acceptedCandidateIds = new Set(acceptedPairs.map((pair) => pair.candidateId));
  for (const pair of requested) {
    const suggestion = suggestionByAsset.get(pair.assetId);
    if (suggestion === undefined || !suggestion.candidateIds.includes(pair.candidateId)) {
      throw new Error("Accepted relink pair is not in the draft.");
    }
    if (acceptedAssetIds.has(pair.assetId)) {
      const existing = acceptedPairs.find((item) => item.assetId === pair.assetId);
      if (existing?.candidateId === pair.candidateId) continue;
      throw new Error("Relink asset is already paired.");
    }
    if (acceptedCandidateIds.has(pair.candidateId)) {
      throw new Error("Relink candidates are one-to-one.");
    }
    if (suggestion.rank === null) throw new Error("Accepted relink pair has no ranked match.");
    acceptedPairs.push({ assetId: pair.assetId, candidateId: pair.candidateId, rank: suggestion.rank });
    acceptedAssetIds.add(pair.assetId);
    acceptedCandidateIds.add(pair.candidateId);
  }
  const unresolvedAssetIds = draft.missingAssets
    .map((asset) => asset.assetId)
    .filter((assetId) => !acceptedAssetIds.has(assetId));
  const unresolvedCandidateCount = draft.candidates.length - acceptedCandidateIds.size;
  return {
    catalogId: draft.catalogId,
    sessionId: draft.sessionId,
    operationId: draft.operationId,
    acceptedPairs,
    unresolvedAssetIds,
    unresolvedAssetCount: unresolvedAssetIds.length,
    unresolvedCandidateCount,
    ambiguousAssetCount: draft.ambiguousAssetCount,
  };
}

export function cancelRelinkDraft(value: unknown): RelinkDraft {
  return planRelink(value);
}

File: src-tauri/src/catalog/relink.rs
use super::*;
use std::collections::{HashMap, HashSet};
use std::sync::atomic::AtomicBool;

pub(super) struct StoredDraft {
    pub(super) public: Value,
    selected: HashMap<String, Value>,
    expires_at: i64,
}

fn filename(path: &str) -> &str {
    path.rsplit('/').next().unwrap_or(path)
}

fn observation(asset: &Value) -> Value {
    if asset["observedAt"].is_null()
        && asset["byteLength"].is_null()
        && asset["modifiedAt"].is_null()
        && asset["localFileId"].is_null()
    {
        Value::Null
    } else {
        json!({"byteLength":asset["byteLength"],"modifiedAt":asset["modifiedAt"],"localFileId":asset["localFileId"]})
    }
}

fn rank(asset: &Value, candidate: &Value) -> Option<&'static str> {
    if asset["fingerprint"]["status"] == "valid"
        && asset["fingerprint"]["sha256"] == candidate["fingerprint"]["sha256"]
        && !asset["fingerprint"]["sha256"].is_null()
    {
        return Some("exact-sha256");
    }
    if !asset["observation"]["localFileId"].is_null()
        && asset["observation"]["localFileId"] == candidate["observation"]["localFileId"]
    {
        return Some("local-file-identity");
    }
    if asset["relativePath"] == candidate["relativePath"]
        && !asset["observation"].is_null()
        && asset["observation"]["byteLength"] == candidate["observation"]["byteLength"]
        && asset["observation"]["modifiedAt"] == candidate["observation"]["modifiedAt"]
    {
        return Some("relative-observation");
    }
    if asset["filename"] == candidate["filename"]
        && !asset["observation"]["byteLength"].is_null()
        && asset["observation"]["byteLength"] == candidate["observation"]["byteLength"]
    {
        return Some("filename-size");
    }
    None
}

fn plan(
    catalog_id: &str,
    session_id: &str,
    operation_id: &str,
    missing: Vec<Value>,
    candidates: Vec<Value>,
) -> Value {
    let mut owners = HashMap::<String, usize>::new();
    for asset in &missing {
        for candidate in &candidates {
            if rank(asset, candidate) == Some("exact-sha256") {
                let id = candidate["candidateId"].as_str().unwrap_or("");
                *owners.entry(id.into()).or_default() += 1;
            }
        }
    }
    let mut preselected = Vec::new();
    let mut used = HashSet::new();
    for asset in &missing {
        let matches = candidates
            .iter()
            .filter(|c| rank(asset, c) == Some("exact-sha256"))
            .collect::<Vec<_>>();
        if matches.len() == 1
            && owners
                .get(matches[0]["candidateId"].as_str().unwrap_or(""))
                .copied()
                == Some(1)
        {
            let candidate_id = matches[0]["candidateId"].as_str().unwrap_or("").to_string();
            used.insert(candidate_id.clone());
            preselected.push(json!({"assetId":asset["assetId"],"candidateId":candidate_id,"rank":"exact-sha256"}));
        }
    }
    let mut suggestions = Vec::new();
    let mut ambiguous = 0usize;
    for asset in &missing {
        if let Some(pair) = preselected
            .iter()
            .find(|p| p["assetId"] == asset["assetId"])
        {
            suggestions.push(json!({"assetId":asset["assetId"],"rank":"exact-sha256","candidateIds":[pair["candidateId"]],"preselectedCandidateId":pair["candidateId"]}));
            continue;
        }
        let exact = candidates
            .iter()
            .filter(|c| rank(asset, c) == Some("exact-sha256"))
            .collect::<Vec<_>>();
        let (selected, match_rank) = if !exact.is_empty() {
            (exact, Some("exact-sha256"))
        } else {
            let mut result = (Vec::new(), None);
            for desired in [
                "local-file-identity",
                "relative-observation",
                "filename-size",
            ] {
                let matches = candidates
                    .iter()
                    .filter(|c| {
                        !used.contains(c["candidateId"].as_str().unwrap_or(""))
                            && rank(asset, c) == Some(desired)
                    })
                    .collect::<Vec<_>>();
                if !matches.is_empty() {
                    result = (matches, Some(desired));
                    break;
                }
            }
            result
        };
        if selected.len() > 1 {
            ambiguous += 1
        }
        suggestions.push(json!({"assetId":asset["assetId"],"rank":match_rank,"candidateIds":selected.iter().map(|v|v["candidateId"].clone()).collect::<Vec<_>>(),"preselectedCandidateId":null}));
    }
    json!({"catalogId":catalog_id,"sessionId":session_id,"operationId":operation_id,"missingAssets":missing,"candidates":candidates,"suggestions":suggestions,"preselectedPairs":preselected,"unresolvedAssetCount":missing.len()-used.len(),"ambiguousAssetCount":ambiguous,"expiresAt":now()+5*60*1000})
}

impl CatalogService {
    pub(super) fn relink_dispatch(&mut self, command: &str, args: &Value) -> Result<Value, String> {
        let input = first(args)?;
        self.require_session(input)?;
        self.relink_drafts
            .retain(|_, draft| draft.expires_at > now());
        match command {
            "darkroom:catalog-relink-files-prepare" => self.relink_prepare(args),
            "darkroom:catalog-relink-files-apply" => self.relink_apply(input),
            "darkroom:catalog-relink-files-cancel" => {
                let id = string(input, "operationId")?;
                self.relink_drafts.remove(id);
                Ok(Value::Null)
            }
            _ => Err("Unsupported relink command.".into()),
        }
    }

    fn relink_prepare(&mut self, args: &Value) -> Result<Value, String> {
        let input = first(args)?;
        let Some(paths) = args
            .as_array()
            .and_then(|v| v.get(1))
            .and_then(Value::as_array)
        else {
            return Ok(Value::Null);
        };
        if paths.is_empty() {
            return Ok(Value::Null);
        }
        if paths.len() > 250 {
            return Err("Relink candidate count is too large.".into());
        }
        if self.relink_drafts.len() >= 32 {
            return Err("Too many relink drafts are active.".into());
        }
        let catalog_id = string(input, "catalogId")?;
        let session_id = string(input, "sessionId")?;
        let roots = rows(
            self.db()?,
            "SELECT root_id AS rootId,canonical_path AS canonicalPath FROM roots WHERE catalog_id=? AND health='online' AND canonical_path IS NOT NULL ORDER BY root_id",
            vec![SqlValue::Text(catalog_id.into())],
        )?;
        if roots.is_empty() {
            return Err("No active catalog roots are available.".into());
        }
        let mut selected = HashMap::new();
        let mut seen = HashSet::new();
        let mut candidates = Vec::new();
        for value in paths {
            let path = value.as_str().ok_or("Relink selection is invalid.")?;
            let metadata =
                fs::symlink_metadata(path).map_err(|_| "Relink candidate is unavailable.")?;
            if !metadata.is_file() || metadata.file_type().is_symlink() {
                return Err("Selected relink item is not a regular file.".into());
            }
            let canonical =
                fs::canonicalize(path).map_err(|_| "Relink candidate is unavailable.")?;
            if !seen.insert(canonical.clone()) {
                return Err("A relink file was selected more than once.".into());
            }
            let matches = roots
                .iter()
                .filter_map(|root| {
                    let base = Path::new(root["canonicalPath"].as_str()?);
                    canonical
                        .strip_prefix(base)
                        .ok()
                        .filter(|relative| !relative.as_os_str().is_empty())
                        .map(|relative| (root, relative))
                })
                .collect::<Vec<_>>();
            if matches.len() != 1 {
                return Err(if matches.is_empty() {
                    "Selected relink file is outside active catalog roots."
                } else {
                    "Selected relink file belongs to multiple active roots."
                }
                .into());
            }
            let (root, relative) = matches[0];
            let relative_path = relative
                .to_str()
                .ok_or("Selected relink file has an invalid relative path.")?
                .replace(std::path::MAIN_SEPARATOR, "/");
            if relative_path
                .split('/')
                .any(|part| part.is_empty() || part == "." || part == "..")
            {
                return Err("Selected relink file has an invalid relative path.".into());
            }
            let observed = fingerprint::file_observation(&metadata);
            let probe =
                json!({"canonicalRootPath":root["canonicalPath"],"relativePath":relative_path});
            let (state, sha, _) = fingerprint::hash(&probe, &AtomicBool::new(false));
            let fingerprint = if state == "indexed" {
                json!({"status":"valid","sha256":sha})
            } else {
                json!({"status":"failed","sha256":null})
            };
            let candidate_id = Uuid::new_v4().to_string();
            selected.insert(candidate_id.clone(),json!({"rootId":root["rootId"],"relativePath":relative_path,"absolutePath":canonical,"observation":observed}));
            candidates.push(json!({"candidateId":candidate_id,"rootId":root["rootId"],"relativePath":relative_path,"filename":filename(&relative_path),"observation":{"byteLength":observed["size"],"modifiedAt":observed["modifiedAt"],"localFileId":observed["localFileId"]},"fingerprint":fingerprint}));
        }
        let raw = rows(
            self.db()?,
            "SELECT a.asset_id AS assetId,a.root_id AS rootId,a.relative_path AS relativePath,a.observed_at AS observedAt,a.observed_byte_length AS byteLength,a.observed_modified_at AS modifiedAt,a.local_file_id AS localFileId,f.status AS fingerprintStatus,f.sha256 AS fingerprintSha256 FROM assets a JOIN fingerprints f ON f.catalog_id=a.catalog_id AND f.asset_id=a.asset_id WHERE a.catalog_id=? AND a.health='missing' ORDER BY a.relative_path,a.asset_id",
            vec![SqlValue::Text(catalog_id.into())],
        )?;
        if raw.len() > 250 {
            return Err("Relink missing asset count is too large.".into());
        }
        let missing=raw.iter().map(|asset|json!({"assetId":asset["assetId"],"rootId":asset["rootId"],"relativePath":asset["relativePath"],"filename":filename(asset["relativePath"].as_str().unwrap_or("")),"observation":observation(asset),"fingerprint":{"status":asset["fingerprintStatus"],"sha256":asset["fingerprintSha256"]}})).collect::<Vec<_>>();
        let operation_id = Uuid::new_v4().to_string();
        let draft = plan(catalog_id, session_id, &operation_id, missing, candidates);
        self.relink_drafts.insert(
            operation_id,
            StoredDraft {
                expires_at: draft["expiresAt"].as_i64().unwrap_or(0),
                public: draft.clone(),
                selected,
            },
        );
        Ok(draft)
    }

    fn relink_apply(&mut self, input: &Value) -> Result<Value, String> {
        let id = string(input, "operationId")?.to_string();
        let draft = self
            .relink_drafts
            .get(&id)
            .ok_or("Relink draft is missing or expired.")?;
        if draft.public["catalogId"] != input["catalogId"]
            || draft.public["sessionId"] != input["sessionId"]
        {
            return Err("Relink draft belongs to a different catalog session.".into());
        }
        let requested = field(input, "acceptedPairs")?
            .as_array()
            .ok_or("Relink accepted pairs are invalid.")?;
        if requested.len() > 250 {
            return Err("Relink accepted pair count is invalid.".into());
        }
        let mut pairs = draft.public["preselectedPairs"]
            .as_array()
            .cloned()
            .unwrap_or_default();
        let mut used_assets = pairs
            .iter()
            .filter_map(|v| v["assetId"].as_str().map(str::to_string))
            .collect::<HashSet<_>>();
        let mut used_candidates = pairs
            .iter()
            .filter_map(|v| v["candidateId"].as_str().map(str::to_string))
            .collect::<HashSet<_>>();
        for pair in requested {
            let asset_id = string(pair, "assetId")?;
            let candidate_id = string(pair, "candidateId")?;
            let suggestion = draft.public["suggestions"]
                .as_array()
                .and_then(|v| v.iter().find(|v| v["assetId"] == asset_id))
                .ok_or("Accepted relink pair is not in the draft.")?;
            if !suggestion["candidateIds"]
                .as_array()
                .is_some_and(|v| v.iter().any(|v| v == candidate_id))
            {
                return Err("Accepted relink pair is not in the draft.".into());
            }
            if used_assets.contains(asset_id) {
                if pairs
                    .iter()
                    .any(|v| v["assetId"] == asset_id && v["candidateId"] == candidate_id)
                {
                    continue;
                }
                return Err("Relink asset is already paired.".into());
            }
            if used_candidates.contains(candidate_id) {
                return Err("Relink candidates are one-to-one.".into());
            }
            used_assets.insert(asset_id.into());
            used_candidates.insert(candidate_id.into());
            pairs.push(
                json!({"assetId":asset_id,"candidateId":candidate_id,"rank":suggestion["rank"]}),
            );
        }
        let draft = self
            .relink_drafts
            .get(&id)
            .ok_or("Relink draft is missing or expired.")?;
        let catalog_id = string(input, "catalogId")?;
        let mut mutations = Vec::new();
        for pair in &pairs {
            let asset_id = string(pair, "assetId")?;
            let candidate_id = string(pair, "candidateId")?;
            let source = one(
                self.db()?,
                "SELECT health FROM assets WHERE catalog_id=? AND asset_id=?",
                values(&[&json!(catalog_id), &json!(asset_id)]),
            )?
            .ok_or("Relink source asset is missing.")?;
            if source["health"] != "missing" {
                return Err("Relink source asset is no longer missing.".into());
            }
            let candidate = draft
                .selected
                .get(candidate_id)
                .ok_or("Relink candidate is not in the draft.")?;
            let root=one(self.db()?,"SELECT canonical_path AS canonicalPath,health FROM roots WHERE catalog_id=? AND root_id=?",values(&[&json!(catalog_id),&candidate["rootId"]]))?.ok_or("Relink destination root is missing.")?;
            if root["health"] != "online" {
                return Err("Relink destination root is unavailable.".into());
            }
            let base = Path::new(string(&root, "canonicalPath")?);
            let path = Path::new(string(candidate, "absolutePath")?);
            if !path.starts_with(base)
                || fs::canonicalize(path)
                    .map_err(|_| "Relink candidate changed or is unavailable.")?
                    != path
            {
                return Err("Relink candidate changed or is unavailable.".into());
            }
            let metadata = fs::symlink_metadata(path)
                .map_err(|_| "Relink candidate changed or is unavailable.")?;
            if !metadata.is_file() || metadata.file_type().is_symlink() {
                return Err("Relink candidate changed or is unavailable.".into());
            }
            let observed = fingerprint::file_observation(&metadata);
            if !fingerprint::same_stat(&observed, &candidate["observation"]) {
                return Err("Relink candidate changed or is unavailable.".into());
            }
            if one(self.db()?,"SELECT 1 FROM assets WHERE catalog_id=? AND root_id=? AND relative_path=? AND asset_id<>?",values(&[&json!(catalog_id),&candidate["rootId"],&candidate["relativePath"],&json!(asset_id)]))?.is_some() { return Err("Relink destination is already occupied.".into()) }
            mutations.push(json!({"kind":"asset-relocate","assetId":asset_id,"rootId":candidate["rootId"],"relativePath":candidate["relativePath"],"observation":{"byteLength":observed["size"],"modifiedAt":observed["modifiedAt"],"observedAt":now(),"localFileId":observed["localFileId"]},"health":"present"}));
        }
        let revision = self.revision(catalog_id)?;
        let applied = if mutations.is_empty() {
            json!({"catalogId":catalog_id,"revision":revision,"changed":false,"appliedMutations":0,"auditId":null})
        } else {
            self.apply(
                &json!({"catalogId":catalog_id,"expectedRevision":revision,"mutations":mutations}),
            )?
        };
        let draft = self
            .relink_drafts
            .remove(&id)
            .ok_or("Relink draft is missing or expired.")?;
        let unresolved = draft.public["missingAssets"]
            .as_array()
            .into_iter()
            .flatten()
            .filter(|v| !used_assets.contains(v["assetId"].as_str().unwrap_or("")))
            .map(|v| v["assetId"].clone())
            .collect::<Vec<_>>();
        let mut result = json!({"catalogId":catalog_id,"sessionId":input["sessionId"],"operationId":id,"acceptedPairs":pairs,"unresolvedAssetIds":unresolved,"unresolvedAssetCount":unresolved.len(),"unresolvedCandidateCount":draft.public["candidates"].as_array().map_or(0,Vec::len)-used_candidates.len(),"ambiguousAssetCount":draft.public["ambiguousAssetCount"]});
        if let Some(object) = applied.as_object() {
            for (key, value) in object {
                result[key] = value.clone();
            }
        }
        Ok(result)
    }
}
